### PR TITLE
Use node.findall if available (docutils 18.x)

### LINKED ISSUE #399: sphinx 5 compatibility
In one of MNE-Python's CIs we're using sphinx 5.0 beta, and seeing this deprecation warning percolating up from docutils:

```
Traceback (most recent call last):
  File "/home/circleci/python_env/lib/python3.8/site-packages/sphinx/events.py", line 94, in emit
    results.append(listener.handler(self.app, *args))
  File "/home/circleci/python_env/lib/python3.8/site-packages/numpydoc/numpydoc.py", line 104, in relabel_references
    for citation_node in doc.traverse(citation):
  File "/home/circleci/python_env/lib/python3.8/site-packages/docutils/nodes.py", line 240, in traverse
    warnings.warn('nodes.Node.traverse() is obsoleted by Node.findall().',
PendingDeprecationWarning: nodes.Node.traverse() is obsoleted by Node.findall().
```

I can probably take a swing at this in the next week or so (?) but would be happy if someone else picked it up too

### PR DESCRIPTION
**Pull request by drammock**
closes #399

### CODE REVIEW

**Comment by jarrodmillman** [comment]
You can fix the linter warnings via:
```
pip install pre-commit
pre-commit run --all-files --show-diff-on-failure --color always
```
Or do
```
 pip install pre-commit
 pre-commit install
```
and it will automatically run when you try to commit.

**Comment by jarrodmillman** [comment]
We will want to make sure we still support `docutils < 0.18` and `sphinx==4`, I think.

**Comment by drammock** [comment]
pre-commit is failing, looks like it mistakes python 3.10 for python 3.1 (!!)  here's the log

<details>

### version information

```
pre-commit version: 2.19.0
git --version: git version 2.25.1
sys.version:
    3.10.4 | packaged by conda-forge | (main, Mar 24 2022, 17:39:04) [GCC 10.3.0]
sys.executable: /opt/miniforge3/envs/numpydoc-dev/bin/python3.1
os.name: posix
sys.platform: linux
```

### error information

```
An unexpected error has occurred: CalledProcessError: command: ('/opt/miniforge3/envs/numpydoc-dev/bin/python3.1', '-mvirtualenv', '/home/drmccloy/.cache/pre-commit/repowbc6hl47/py_env-python3.1', '-p', 'python3.1')
return code: 1
expected return code: 0
stdout:
    RuntimeError: failed to find interpreter for Builtin discover of python_spec='python3.1'
    
stderr: (none)
```

```
Traceback (most recent call last):
  File "/opt/miniforge3/envs/numpydoc-dev/lib/python3.10/site-packages/pre_commit/error_handler.py", line 73, in error_handler
    yield
  File "/opt/miniforge3/envs/numpydoc-dev/lib/python3.10/site-packages/pre_commit/main.py", line 361, in main
    return hook_impl(
  File "/opt/miniforge3/envs/numpydoc-dev/lib/python3.10/site-packages/pre_commit/commands/hook_impl.py", line 238, in hook_impl
    return retv | run(config, store, ns)
  File "/opt/miniforge3/envs/numpydoc-dev/lib/python3.10/site-packages/pre_commit/commands/run.py", line 414, in run
    install_hook_envs(to_install, store)
  File "/opt/miniforge3/envs/numpydoc-dev/lib/python3.10/site-packages/pre_commit/repository.py", line 223, in install_hook_envs
    _hook_install(hook)
  File "/opt/miniforge3/envs/numpydoc-dev/lib/python3.10/site-packages/pre_commit/repository.py", line 79, in _hook_install
    lang.install_environment(
  File "/opt/miniforge3/envs/numpydoc-dev/lib/python3.10/site-packages/pre_commit/languages/python.py", line 219, in install_environment
    cmd_output_b(*venv_cmd, cwd='/')
  File "/opt/miniforge3/envs/numpydoc-dev/lib/python3.10/site-packages/pre_commit/util.py", line 146, in cmd_output_b
    raise CalledProcessError(returncode, cmd, retcode, stdout_b, stderr_b)
pre_commit.util.CalledProcessError: command: ('/opt/miniforge3/envs/numpydoc-dev/bin/python3.1', '-mvirtualenv', '/home/drmccloy/.cache/pre-commit/repowbc6hl47/py_env-python3.1', '-p', 'python3.1')
return code: 1
expected return code: 0
stdout:
    RuntimeError: failed to find interpreter for Builtin discover of python_spec='python3.1'
    
stderr: (none)
```

</details>

**Comment by drammock** [comment]
> We will want to make sure we still support `docutils < 0.18` and `sphinx==4`, I think.

my change to `test_reference` will break that, I think.  Stand by.

File: numpydoc/tests/test_full.py
Comment (jarrodmillman) [inline numpydoc/tests/test_full.py:99]:
```suggestion
    # TODO: This check can be removed when the minimum supported docutils version
    # for numpydoc is docutils>=0.18
    pattern = (
        'role="doc-backlink"'
        if version.parse(docutils_version) >= version.parse("0.18")
        else 'class="fn-backref"'
    )
    reference_list = re.findall(rf'<a {pattern} href="\#id\d+">(.*)<\/a>', html)
```

**Review by jarrodmillman — CHANGES_REQUESTED** [review]

File: numpydoc/numpydoc.py
Comment (drammock) [reply numpydoc/numpydoc.py:47]:
```suggestion
    """Triage node.traverse (docutils <0.18) vs node.findall.

        TODO: This check can be removed when the minimum supported docutils version
        for numpydoc is docutils>=0.18
    """
```

File: numpydoc/tests/test_full.py
Comment (drammock) [inline numpydoc/tests/test_full.py:98]: This version check should stay as `0.18` though; the following change came in 0.18b1

```
  - Use ARIA roles to enable accessible HTML for abstract, dedication,
    the table of contents, footnote, references, footnotes, citations,
    and backlinks.
```

**Comment by drammock** [comment]
Just a note that although the CIs were coming back green before the last commit, I don't think they're actually testing against sphinx 5.0beta / docutils 0.18 ?  Locally I still see this:

```
=================================== short test summary info ===================================
XFAIL numpydoc/tests/test_validate.py::TestValidator::test_bad_docstrings[BadParameters-blank_lines-msgs30]
XFAIL numpydoc/tests/test_validate.py::TestValidator::test_bad_docstrings[BadReturns-no_type-msgs33]
=============================== 237 passed, 2 xfailed in 1.73s ================================
```
and I'm not sure how to fix it...  that `pytest.parametrize` is over 200 lines long and I haven't had the time to wrap my head around it.  Maybe @datapythonista or @larsoner have time to look?

**Comment by drammock** [comment]
> Just a note that although the CIs were coming back green before the last commit, I don't think they're actually testing against sphinx 5.0beta / docutils 0.18 ? 

Never mind, I'm wrong, I just spotted the `prerelease` jobs.  And I also just realized that I was mis-reading the pytest output.  Those are *successfully failing* XFAILs not *unexpectedly passing* XFAILs.  So I think this is ready for review/merge @jarrodmillman

**Review by larsoner — APPROVED** [review]
Code changes LGTM, +1 for merge. Feel free to merge if you're happy @jarrodmillman !

I can also confirm that on docutils 0.18.1 I have failures building MNE-Python's doc, and with this PR things build fine.

**Comment by jarrodmillman** [comment]
See #404. That should fix the test failures.

**Comment by drammock** [comment]
> See #404. That should fix the test failures.

Ok great! Can you manually re-trigger the failing runs? I can't just push an empty commit; as a first time contrib here, the CIs will wait for maintainer approval anyway.

**Comment by larsoner** [comment]
Since it's a change to the CI yaml you'll need to merge or rebase I think

**Comment by drammock** [comment]
> Since it's a change to the CI yaml you'll need to merge or rebase I think

Ah right, of course. Will do when I'm not on a phone

**Review by jarrodmillman — APPROVED** [review]

**Comment by jarrodmillman** [comment]
Thanks!!

**Pull request merged**
